app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.ts

```diff
@@ -0,0 +1,265 @@
+      ...args: Parameters<AbstractPermissionController['grantPermissions']>
+    ) => Record<string, ValidPermission<string, Caveat<string, Json>>>;
+    requestPermissionApprovalForOrigin: (
+      requestedPermissions: RequestedPermissions,
+    ) => Promise<{ approvedAccounts: Hex[]; approvedChainIds: Hex[] }>;
+
+    getPermissionsForOrigin: () => ReturnType<
+      AbstractPermissionController['getPermissions']
+    >;
+    getAccounts: () => string[];
+  },
+) {
+  const { origin, params } = req;
+
+  if (!Array.isArray(params) || !isPlainObject(params[0])) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  const [requestedPermissions] = params;
+  delete requestedPermissions[Caip25EndowmentPermissionName];
+
+  const legacyRequestedPermissions: Partial<
+    Pick<RequestedPermissions, 'eth_accounts' | 'endowment:permitted-chains'>
+  > = pick(requestedPermissions, [
+    RestrictedMethods.eth_accounts,
+    PermissionNames.permittedChains,
+  ]);
+  delete requestedPermissions[RestrictedMethods.eth_accounts];
+  delete requestedPermissions[PermissionNames.permittedChains];
+
+  // We manually handle eth_accounts and permittedChains permissions
+  // by calling the ApprovalController rather than the PermissionController
+  // because these two permissions do not actually exist in the Permssion
+  // Specifications. Calling the PermissionController with them will
+  // cause an error to be thrown. Instead, we will use the approval result
+  // from the ApprovalController to form a CAIP-25 permission later.
+  let legacyApproval;
+  const haveLegacyPermissions =
+    Object.keys(legacyRequestedPermissions).length > 0;
+  if (haveLegacyPermissions) {
+    if (!legacyRequestedPermissions[RestrictedMethods.eth_accounts]) {
+      legacyRequestedPermissions[RestrictedMethods.eth_accounts] = {};
+    }
+
+    if (!legacyRequestedPermissions[PermissionNames.permittedChains]) {
+      legacyRequestedPermissions[PermissionNames.permittedChains] = {};
+    }
+
+    if (isSnapId(origin)) {
+      delete legacyRequestedPermissions[PermissionNames.permittedChains];
+    }
+
+    legacyApproval = await requestPermissionApprovalForOrigin(
+      legacyRequestedPermissions,
+    );
+  }
+
+  let grantedPermissions: GrantedPermissions = {};
+  // Request permissions from the PermissionController for any permissions other
+  // than eth_accounts and permittedChains in the params. If no permissions
+  // are in the params, then request empty permissions from the PermissionController
+  // to get an appropriate error to be returned to the dapp.
+  if (
+    (Object.keys(requestedPermissions).length === 0 &&
+      !haveLegacyPermissions) ||
+    Object.keys(requestedPermissions).length > 0
+  ) {
+    const [_grantedPermissions] = await requestPermissionsForOrigin(
+      requestedPermissions,
+    );
+    // permissions are frozen and must be cloned before modified
+    grantedPermissions = { ..._grantedPermissions };
```

Comment:
We typically use `_` for unused variables, but this is being used.

Maybe this could be called `frozenGrantedPermissions`?

``` suggestion
    const [frozenGrantedPermissions] = await requestPermissionsForOrigin(
      requestedPermissions,
    );
    // permissions are frozen and must be cloned before modified
    grantedPermissions = { ...frozenGrantedPermissions };
```